### PR TITLE
feat: adds @deepkit/type

### LINKED ISSUE #894: Add deepkit
As a part of the Deepkit framework they provide a runtime type implementation documented here: https://deepkit.io/documentation/type

### PR DESCRIPTION
**Pull request by BasixKOR**
Making a draft since I had a hard time running this on my local machine. Aims to resolve #894.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -22,6 +22,9 @@
   "dependencies": {
     "@ailabs/ts-utils": "1.4.0",
     "@badrap/valita": "0.1.3",
+    "@deepkit/core": "^1.0.1-alpha.65",
+    "@deepkit/type": "^1.0.1-alpha.71",
+    "@deepkit/type-compiler": "^1.0.1-alpha.71",
```
Comment (moltar) [inline package.json:27]: Please use fixed versions. Renovate Bot will update them.

File: cases/deepkit.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+interface DataTypeSafe extends DataType {
+  [key: string]: unknown;
+  deeplyNested: {
+    foo: string;
+    num: number;
+    bool: boolean;
+    [key: string]: unknown;
+  };
+}
```
Comment (moltar) [inline cases/deepkit.ts:26]: @BasixKOR The spirit of this repo is to define types in code, and then infer the types from it AND validate.

If you have to define the interfaces by hand, then it is not suitable for this repository.

Comment (BasixKOR) [reply cases/deepkit.ts:26]: IIRC deepkit uses this interface to generate the runtime types for validation. This is in fact a type definition to be used for validation.

Comment (hoeck) [reply cases/deepkit.ts:26]: @BasixKOR that is exactly how it works and advanced runtime reflection is one big selling point of deepkit.

@moltar IMHO it follows the spirit of "you only have to define your types once" - interface and validator are always in sync.

File: tsconfig.json [hunk 3]
```diff
@@ -7,6 +7,7 @@
     "emitDecoratorMetadata": true,
     "resolveJsonModule": true,
```
Comment (hoeck) [inline tsconfig.json:8]: That trailing comma trips parsing of tsconfig inside deepkit:

```
Could not parse /home/runner/work/typescript-runtime-type-benchmarks/typescript-runtime-type-benchmarks/tsconfig.json: SyntaxError: Unexpected token } in JSON at position 223
```

Comment (BasixKOR) [reply tsconfig.json:8]: Ooh I was wondering what was causing for like forever, thanks for catching

**Comment by hoeck** [comment]
Also, deepkit seems to require at least node v14 (tried that locally and it worked):

```
    /home/runner/work/typescript-runtime-type-benchmarks/typescript-runtime-type-benchmarks/node_modules/@deepkit/type/dist/cjs/src/reflection/reflection.js:923
                type.visibility = prop.visibility ?? type_1.ReflectionVisibility.public;
                                                   ^
    SyntaxError: Unexpected token '?'
      at Runtime.createScriptFromCode (node_modules/jest-runtime/build/index.js:1728:14)
      at Object.<anonymous> (node_modules/@deepkit/type/dist/cjs/src/decorator.js:4:28)
```

To resolve this, we could:

1. add a node version filter to the benchmarks
2. use some tool (babel?) to transpile node-modules
3. drop node 12 benchmarks

In absence of other options, I'd vote for dropping node 12 benchmarks :grin:

**Comment by hoeck** [comment]
@BasixKOR I you like, you can also run the tests locally with `npm run test`, you'll then see that the `parseStrict` case test is failing for deepkit:

```
    deepkit
      ✓ should validate the data (2 ms)
      ✓ should throw on invalid attribute type (1 ms)
      ✕ should throw on extra attributes
      ✕ should throw on extra nested attributes
      ✕ should throw on missing attributes
```

If there is no strict validation option for deepkit (I did not found any, tried `validatedDeserialize()` but that did ignore extra attributes), you can omit the case and only include those that deepkit supports.

File: cases/deepkit.ts (hunk 2 above)
Comment (hoeck) [reply cases/deepkit.ts:26]: @BasixKOR what I don't get tough is `DataTypeSafe` interface - that does not look like it matches the test type at all - should you always just use the `DataType` or maybe `DataType` with `[key: string]: unknown;` to "ignore" additional keys?

**Comment by BasixKOR** [comment]
> use some tool (babel?) to transpile node-modules

@hoeck Yeah, It seems like Jest allows to do this with `transformIgnorePatterns` so I think I'll try this option.

> you can omit the case and only include those that deepkit supports.

Thanks, I'll remove loose and safe test cases for now until I figure out how to ignore unknown keys.

File: cases/deepkit.ts (hunk 2 above)
Comment (BasixKOR) [reply cases/deepkit.ts]: The latter one doesn't really corresponds to the test type too, so I think it's better to test against `strict` cases for now.

**Comment by hoeck** [comment]
> I understood in a completely opposite way joy I'll remove the strict ones, thanks for the suggestion!

Indeed that was worded quite ambiguously, I meant removing the benchmark case :sweat_smile: not the test case.

> Yeah, It seems like Jest allows to do this with transformIgnorePatterns so I think I'll try this option.

That will help with running the Jest tests but you'd still need a separate setup for running the benchmark?

As the [oldest LTS release is node 14 now](https://nodejs.org/en/about/releases/), I am all in favor of dropping node 12 support instead of introducing more build complexity. @moltar any opinions?

File: cases/deepkit.ts (hunk 2 above)
Comment (moltar) [reply cases/deepkit.ts]: Ah, great, didn't realize that. Sorry 😅

**Comment by moltar** [comment]
I am in favour of dropping Node v12!

File: cases/deepkit.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    bool: boolean;
+  };
+}
+
+addCase('deepkit', 'parseSafe', data => {
+  return deserialize<DataType>(data);
+});
+
+addCase('deepkit', 'assertLoose', data => {
+  const result = validate<DataType>(data);
```
Comment (marcj) [inline cases/deepkit.ts:23]: To get a validator for a specific type, you could do

```typescript
const dataTypeGuard = guard<DataType>();

addCase('deepkit', 'assertLoose', data => {
    const valid = dataTypeGuard(data);
});
```

File: cases/deepkit.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+  boolean: boolean;
+  deeplyNested: {
+    foo: string;
+    num: number;
+    bool: boolean;
+  };
+}
+
+addCase('deepkit', 'parseSafe', data => {
+  return deserialize<DataType>(data);
```
Comment (marcj) [inline cases/deepkit.ts:19]: To get a deserializer for a specific type, you could do. Saves a lot of type resolving logic.

```typescript
const dataTypeDeserialize = deserializeFunction<DataType>();

addCase('deepkit', 'parseSafe', data => {
   return dataTypeDeserialize(data);
});
```

**Comment by hoeck** [comment]
> I am in favour of dropping Node v12!

:+1: okay I'll create a PR that drops node 12, after merging, this PR can then be rebased and deepkit tests and benchmarks should work :grinning:

**Comment by hoeck** [comment]
@BasixKOR could you please rebase on the latest master - I have removed node 12, the benchmarks now run on 14, 16 and node 18 only so your PR should pass after the rebase. If you don't have time any more, please let me know and I will take care of this PR. Thanks!

**Comment by BasixKOR** [comment]
I rebased the deepkit branch and it should be okay, please lmk if anything is a bit off

File: cases/deepkit.ts (hunk 5 above)
Comment (hoeck) [reply cases/deepkit.ts:19]: @BasixKOR what do you think about this suggestion from marcj?

**Comment by BasixKOR** [comment]
Fixed some remaining rebase issues and uses guard and deserializeFunction now! Unfortunately 'should throw on missing attributes' test is breaking though.

File: cases/deepkit.ts (hunk 5 above)
Comment (marcj) [reply cases/deepkit.ts]: It has to be done this way otherwise it's not a proper comparison with other cases.

**Pull request closed without merging**